src/sqlparser/src/parser.rs

```diff
@@ -435,9 +435,24 @@ impl Parser {
                 } else {
                     UnaryOperator::Minus
                 };
+                let mut sub_expr = self.parse_subexpr(Self::PLUS_MINUS_PREC)?;
+                // TODO: Deal with nested unary exp: -(-(-(1))) => -1
+                // Tracked by: <https://github.com/singularity-data/risingwave/issues/4344>
+                match sub_expr {
+                    Expr::Value(Value::Number(ref mut s, _))
+                    | Expr::TypedString {
```

Comment:
Why do we need to handle `TypedString` here?

Reply:
Used it to handle `Float`s. Let me verify if this is really needed.
If it is, I will match on datatype as well, so ensure non-numeric typed strings are not included in this pass.

Reply:
Actually not needed for floats. Only needed for integrals:
```rs
fn main() {
    println!("{}", f64::MAX == -f64::MIN);
    println!("{}", f32::MAX == -f32::MIN);
}
```

output:
```
true
true
```

Reply:
Will remove this case, thanks for pointing it out :grin:

Reply:
IIUC
* unary negative followed by typed string means `- smallint '32768'`, which also reports out-of-range error in pg.
* "floats", or `decimals` to be accurate (`-3.5`), are also handled by `Number` case above.